Modified is_positive for cosh function

rohitrango opened the pull request:
Fixes #11721.

`cosh(x).is_positive` returns `True` if x is real.
Also modified the `Abs` function to return the function itself or its negative if the function is positive or negative.

Example:
```
In [1]: r=symbols('r',real=True)

In [2]: abs(cosh(x))
Out[2]: │cosh(x)│

In [3]: abs(cosh(r))
Out[3]: cosh(r)

In [4]: abs(cosh(r)) == cosh(r)
Out[4]: True

In [5]: abs(cosh(x)) == cosh(x)
Out[5]: False

In [6]: cosh(r).is_positive
Out[6]: True

In [7]: cosh(x).is_positive

In [8]:      
```

rohitrango committed:
Updated the `is_positive` property of cosh.
Also modified the Abs function for general real functions

rohitrango committed:
Added tests for cosh(x) function for real x
Also checks if abs(cosh(x)) returns cosh(x) for real x.

rohitrango commented:
@asmeurer I think this should fix the given issue. Thank you.

jksuom commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -209,6 +209,11 @@ class cosh(HyperbolicFunction):
 
+    @property
+    def is_positive(self):
+        if self._eval_is_real():
```
This should not be based on reality, since cosh also takes negative real values.

rohitrango replied:
@jksuom, this is the definition of the _eval_is_real() function-

def _eval_is_real(self):
        return self.args[0].is_real

So basically it checks whether the parameter passed (in this case, x) is real, and not whether the function cosh is real valued.

rohitrango replied:
Or am I still missing something?

jksuom replied:
We should not rely on the implementation of `_eval_is_real()` (which is currently wrong).

jksuom replied:
If that is ever extended to include all purely imaginary arguments, then this implementation would be wrong.

rohitrango replied:
By the way, shouldn't the method `_eval_is_real()` be always true? Because for a general number `a + ib` (a,b are real), `cosh(a + ib) = exp(a)*cos(b)` which is always real.

rohitrango replied:
Oops, my bad! I did a silly mistake. But the test isn't that difficult either, should I create an issue for that?

rohitrango replied:
Alright! Will do. Can you please re-run the Travis build? All tests are working fine on my computer.

jksuom replied:
Sphinx test is failing. It hard to tell why. You should test that locally. It seems that restart will not help.

rohitrango committed:
Condition for is_positive corrected

jksuom commented on sympy/functions/elementary/complexes.py:
```diff
@@ -503,6 +503,16 @@ def eval(cls, arg):
 
+        # if the expression is a function, then try to see if its positive or negative
+        # functions which have `is_positive` and `is_negative` values of form - bool or None
+        if isinstance(arg, Function) and not isinstance(arg.is_positive,property) \
```
The second argument of `isinstance()` should have classinfo. 'property' is not a class.

rohitrango committed:
Modified is_nonnegative for cosh function.
Also modified the `Abs` function.

Some functions have incorrect (or incomplete) implementation of
`is_positive`, `is_nonnegative` properties.
Hence, the modified conditions.

rohitrango commented on sympy/functions/elementary/complexes.py:
```diff
@@ -503,6 +503,15 @@ def eval(cls, arg):
+        # if the expression is a function, then try to see if its nonnegative or negative
+        # Some class of Functions like UndefinedFunctions return a `property` function
+        if isinstance(arg, Function):
+            if arg.is_nonnegative is True or arg.is_positive is True:
```
Oh. The reason it didn't work earlier is because `is_nonnegative` and `is_positive` wasn't implemented for `cosh`. 
Still, it didn't work because I only implemented `is_positive` for `cosh`.
I'm removing that case then.

jksuom replied:
That Sphinx error is puzzling. Maybe its version should be updated.

rohitrango commented:
I think this should be reviewed now. I'm sorry for the delayed response, had some other work piled up which needed my attention.

I'm looking into the other PRs asap.

jksuom commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -209,6 +209,16 @@ class cosh(HyperbolicFunction):
+    @property
+    def is_positive(self):
+        if self.args[0].is_real:
+            return True
```
This is wrong since the value can also be negative.

jksuom replied:
Correction: That is not exactly wrong but there are other (non-real) argument values for which cosh is positive or negative. Perhaps this should include all of them?

gxyd replied:
But wouldn't that mean that we will be trying to solve inequality in complex domain? Which is neither supported in SymPy nor will make much sense. The inequality will be `t**2 + 1 > 0` where `t = exp(z)`.

rohitrango replied:
@gxyd I didn't really get that. Could you explain in a bit more detail ?

jksuom replied:
It is not necessary to solve inequalities (in the real or complex domain), since we have explicit formulas for the real and imaginary parts of cosh(x + iy) for real x and y.

gxyd replied:
~~Consider that `cosh(z).is_positive` where `z` is a complex number, that would be equivalent to solving the complex inequality `solveset(cosh(z) > 0, z)` (try this in SymPy) gives "inequalities not supported in complex domain".~~
I think @jksuom is right :)

rohitrango replied:
Yes, the checks are made by taking those forms of the expression as @jksuom said.

rohitrango committed:
Added more cases of `is_positive` and `is_nonnegative`
for complex arguments.

jksuom commented:
It might be more efficient to avoid introducing  cos(im) when we know that im is real. Instead we could use  `im % (2*pi) < pi/2 or 3*pi/2 < im % (2*pi)`.

jksuom commented:
I am a little unsure about the behaviour of SymPy booleans but I suspect that it could even be possible to return something like `Or(im % (2*pi) < pi/2 , 3*pi/2 < im % (2*pi))`. That would contain the essential information also when it cannot be evaluated but remains in symbolic form. (Some testing would be needed.)

jksuom commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -209,6 +209,52 @@ class cosh(HyperbolicFunction):
+            return True
+
+        re, im = arg.as_real_imag()
+        if im % pi == 0 or re == 0:
```
Would `im % (2*pi) == 0` suffice here?

rohitrango replied:
cosh(i*pi/4) is positive.

jksuom replied:
That should be taken care of by the other condition: re == 0.

jksuom replied:
I wanted to say: `2*pi` instead of pi. That would make cos(im) positive.

rohitrango replied:
Yes, that can be done, but I don't think there would be a big performance difference due to that. I think the current way works fine as well.

rohitrango replied:
Oh, I get it. Yes, this is good. Is there anything else I can improve?

jksuom replied:
I have not much experience with boolean expressions but maybe something like this would work:
```
x, y = arg.as_real_imag()
yred = y % (2*pi)
return yred == 0 | x == 0 & ((yred < pi/2) | (3*pi/2 < yred))
```
(I'm not sure if there should be additional parentheses.)

jksuom replied:
I think your observations are correct.

I was hoping that we could return a symbolic expression in case of non-numbers. I still think that should be possible. What would happen if `x == 0` were replaced with `Eq(x, 0)`?

As to [12] above, the test should be `if Or(k < 1, k > 3) is True`, since almost anything nonempty will count as True. (And similarly with `(k < 1) | (k > 3)`.)

jksuom replied:
Yes, I think that | will be internally translated to Or.

rohitrango replied:
Yes.

gxyd commented:
@rohitrango would you like address the commented changes?

rohitrango commented:
@gxyd I'm sorry I didn't get you. Do I have any more changes to make here?

gxyd commented:
I was having an impression from the conversation (from comments by @jksuom ) that it could be possible to refactor your changes to return compound boolean expression. 

@jksuom do you think changes can (or should) be made?

BTW, thanks for responding quickly :)

rohitrango commented:
@gxyd I thought that these were the final changes, and hence, I didn't come back to address these issues (or get them merged or closed). My sincere apologies. I think I will wait for @jksuom to suggest any changes before I work on them.

jksuom commented:
I think that returning a boolean expression instead of `None` could be preferable in some situations where the inequalities raise `TypeError`. It would be good to have more unit tests for such cases before making final decisions.

rohitrango committed:
Modified the functions to handle simple cases first.

gxyd commented on sympy/functions/elementary/complexes.py:
```diff
@@ -482,9 +482,9 @@ def eval(cls, arg):
         if arg.is_zero:
             return S.Zero
-        if arg.is_nonnegative:
+        if arg.is_nonnegative or arg.is_positive:
```
If `arg.is_positive` is True then wouldn't `arg.is_nonnegative` be also True?

rohitrango replied:
Yes, but I guess I had put it just in case it wasn't taken care of in some other class.

gxyd replied:
We should probably remove this. Unless you have an example/logic that justified having this.

gxyd commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -209,6 +209,60 @@ class cosh(HyperbolicFunction):
 
+    @property
+    def is_positive(self):
+        from sympy import pi, Eq
```
Can we have these (infact all, including the ones below) as top-level imports?

rohitrango replied:
Okay. I think I messed it up. What could have gone wrong here? I'll try to fix the errors.

gxyd replied:
Can you have this `import` statement at the top of the file?

rohitrango replied:
I made both the changes, but I'm still failing a few cases. I'm not sure what is causing the errors.

gxyd commented:
There are still some valid tests failures.

rohitrango committed:
Modified import locations

symbolique commented:
You probably shouldn't define is_positive directly, but rather define _eval_is_positive to make the result go through the assumption system.

oscarbenjamin commented:
The referenced issue is now closed due to #13678. Is this PR still needed?

smichr commented:
> Is this PR still needed?

that issue commented, "This also fixes issue #11721, but not as thoroughly as pull request #11924."

czgdp1807 commented:
@rohitrango It looks like a good PR. Can we continue this?
